contracts/staking/TokenStaking.sol

```diff
@@ -0,0 +1,421 @@
+        uint256 deallocationDuration;
+        uint256 minAllocationSize;
+    }
+
+    T public immutable token;
+    IKeepTokenStaking public immutable keepStakingContract;
+    INuCypherStakingEscrow public immutable nucypherStakingContract;
+    VendingMachine public immutable keepVendingMachine;
+    VendingMachine public immutable nucypherVendingMachine;
+
+    mapping(address => uint256) public tStakers;
+
+    mapping(address => mapping(StakingProvider => mapping(address => AppAllocation)))
+        public allocationsPerStaker;
+    // TODO public getter for availability
+    mapping(address => ApplicationInfo) public appInfo;
+    address[] public apps;
+
+    /// @param _token Address of T token contract
+    /// @param _keepStakingContract Address of Keep staking contract
+    /// @param _nucypherStakingContract Address of NuCypher staking contract
+    /// @param _keepVendingMachine Address of Keep vending machine
+    /// @param _nucypherVendingMachine Address of NuCypher vending machine
+    constructor(
+        T _token,
+        IKeepTokenStaking _keepStakingContract,
+        INuCypherStakingEscrow _nucypherStakingContract,
+        VendingMachine _keepVendingMachine,
+        VendingMachine _nucypherVendingMachine
+    ) {
+        // TODO check contracts and input variables
+        token = _token;
+        keepStakingContract = _keepStakingContract;
+        nucypherStakingContract = _nucypherStakingContract;
+        keepVendingMachine = _keepVendingMachine;
+        nucypherVendingMachine = _nucypherVendingMachine;
+    }
+
+    /// @notice Penalizes staker `staker`; the penalty details are encoded in `penaltyData`
+    function slashStaker(address staker, bytes calldata penaltyData) external {
+        // TODO check slashing event (from app)
+        // TODO choose stake (and check allocation)
+        // TODO prepare caldata
+        // TODO send slashing call
+    }
+
+    /// @notice Enable/disable application for the specified staking providers and reset deallocation duration
+    function setupApplication(
+        address app,
+        StakingProvider[] calldata stakingProviders,
+        bool availability
+    ) external onlyOwner {
+        ApplicationInfo storage info = appInfo[app];
+        for (uint256 i = 0; i < stakingProviders.length; i++) {
+            StakingProvider stakingProvider = stakingProviders[i];
+            info.availability[stakingProvider] = availability;
+        }
+        info.deallocationDuration = IApplication(app).deallocationDuration();
+        info.minAllocationSize = IApplication(app).minAllocationSize();
+
+        bool existingApp = false;
+        for (uint256 i = 0; i < apps.length; i++) {
+            if (apps[i] == app) {
+                existingApp = true;
+                break;
+            }
+        }
+        if (!existingApp) {
+            apps.push(app);
+        }
+    }
+
+    /// @notice Depost and stake T tokens
+    function stake(uint256 value) external {
```

Comment:
I'd love to maintain the same roles as in the Keep staking contract, if possible: 
- owner
- operator
- beneficiary
- authorizer

I think this setup proved to be successful on mainnet and Keep community is used to it.